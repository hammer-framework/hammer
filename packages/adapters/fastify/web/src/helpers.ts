/** Ensures that `path` starts and ends with a slash ('/') */
export function coerceRootPath(path: string) {
  const prefix = path.charAt(0) !== '/' ? '/' : ''
  const suffix = path.charAt(path.length - 1) !== '/' ? '/' : ''

  return `${prefix}${path}${suffix}`
}
