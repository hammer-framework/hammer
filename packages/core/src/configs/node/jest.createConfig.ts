const path = require('path')

export default function getNodeJestConfig() {
  return {
    displayName: {
      color: 'redBright',
      name: 'api',
    },

    resolver: 'jest-directory-named-resolver',
    setupFilesAfterEnv: [path.resolve(__dirname, './jest.setup.js')],
  }
}
