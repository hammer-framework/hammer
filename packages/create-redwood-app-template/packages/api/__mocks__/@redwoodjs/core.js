import path from 'path'

const BASE_PATH = path.resolve(__dirname, '../../src/__tests__/fixtures')

export const getPaths = () => ({
  api: {
    services: path.resolve(BASE_PATH, './api/src/services'),
    graphql: path.resolve(BASE_PATH, './api/src/graphql'),
  },
})
