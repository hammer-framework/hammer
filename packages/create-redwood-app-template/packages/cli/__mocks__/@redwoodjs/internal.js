import path from 'path'

const BASE_PATH = '.'

export const getPaths = () => ({
  base: BASE_PATH,
  api: {
    src: path.resolve(BASE_PATH, './api/src'),
    services: path.resolve(BASE_PATH, './api/src/services'),
    graphql: path.resolve(BASE_PATH, './api/src/graphql'),
  },
  web: {
    routes: './web/src/Routes.js',
  },
})
