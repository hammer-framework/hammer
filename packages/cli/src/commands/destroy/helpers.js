import Listr from 'listr'

import { deleteFilesTask } from 'src/lib'
import c from 'src/lib/colors'

const tasks = ({ componentName, filesFn, name }) =>
  new Listr(
    [
      {
        title: `Destroying ${componentName} files...`,
        task: async () => {
          const f = await filesFn({ name })
          return deleteFilesTask(f)
        },
      },
    ],
    { collapse: false, exitOnError: true }
  )

export const createYargsForComponentDestroy = ({ componentName, filesFn }) => {
  return {
    command: `${componentName} <name>`,
    desc: `Destroy a ${componentName} component.`,
    handler: async ({ name }) => {
      const t = tasks({ componentName, filesFn, name })

      try {
        await t.run()
      } catch (e) {
        console.log(c.error(e.message))
      }
    },
    tasks,
  }
}
