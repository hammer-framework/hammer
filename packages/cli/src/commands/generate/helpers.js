import path from 'path'

import camelcase from 'camelcase'
import pluralize from 'pluralize'
import Listr from 'listr'
import pascalcase from 'pascalcase'
import { paramCase } from 'param-case'

import { generateTemplate, getPaths, writeFilesTask } from 'src/lib'
import c from 'src/lib/colors'

/**
 * Reduces boilerplate for generating an output path and content to write to disk
 * for a component.
 */
// TODO: Make this read all the files in a template directory instead of
// manually passing in each file.
export const templateForComponentFile = ({
  name,
  suffix = '',
  extension = '.js',
  webPathSection,
  apiPathSection,
  generator,
  templatePath,
  templateVars,
  componentName,
}) => {
  const basePath = webPathSection
    ? getPaths().web[webPathSection]
    : getPaths().api[apiPathSection]
  const outputComponentName =
    componentName || pascalcase(paramCase(name)) + suffix
  const outputPath = path.join(
    basePath,
    outputComponentName,
    outputComponentName + extension
  )
  const content = generateTemplate(
    path.join(generator, 'templates', templatePath),
    {
      name,
      outputPath: `./${path.relative(getPaths().base, outputPath)}`,
      ...templateVars,
    }
  )
  return [outputPath, content]
}

/**
 * Creates a route path, either returning the existing path if passed, otherwise
 * creates one based on the name
 */
export const pathName = (path, name) => {
  return path ?? `/${paramCase(name)}`
}

/**
 * Reduces boilerplate for creating a yargs handler that writes a component to a
 * location.
 */
// TODO: Make this work for all files, not just components.
export const createYargsForComponentGeneration = ({
  componentName,
  filesFn,
}) => {
  return {
    command: `${componentName} <name>`,
    desc: `Generate a ${componentName} component.`,
    builder: { force: { type: 'boolean', default: false } },
    handler: async ({ force, ...rest }) => {
      const tasks = new Listr(
        [
          {
            title: `Generating ${componentName} files...`,
            task: async () => {
              const f = await filesFn(rest)
              return writeFilesTask(f, { overwriteExisting: force })
            },
          },
        ],
        { collapse: false, exitOnError: true }
      )

      try {
        await tasks.run()
      } catch (e) {
        console.log(c.error(e.message))
      }
    },
  }
}

// Returns all relations to other models
export const relationsForModel = (model) => {
  return model.fields
    .filter((f) => f.relationName)
    .map((field) => {
      const relationName = camelcase(field.type)
      return field.isList ? pluralize(relationName) : relationName
    })
}

// Returns only relations that are of datatype Int
export const intForeignKeysForModel = (model) => {
  return model.fields
    .filter((f) => f.name.match(/Id$/) && f.type === 'Int')
    .map((f) => f.name)
}
