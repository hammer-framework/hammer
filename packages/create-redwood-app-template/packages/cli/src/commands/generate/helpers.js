import path from 'path'

import Listr from 'listr'
import pascalcase from 'pascalcase'

import { generateTemplate, getPaths, writeFilesTask } from 'src/lib'

/**
 * Reduces boilerplate for generating an output path and content to write to disk
 * for a component.
 */
// TODO: Make this read all the files in a template directory instead of
// manually passing in each file.
export const templateForComponentFile = ({
  name,
  suffix = '',
  extension = '.js',
  webPathSection,
  templatePath,
  templateVars,
}) => {
  const basePath = getPaths().web[webPathSection]
  const componentName = pascalcase(name) + suffix
  const outputPath = path.join(
    basePath,
    componentName,
    componentName + extension
  )
  const content = generateTemplate(templatePath, {
    name,
    outputPath: `./${path.relative(getPaths().base, outputPath)}`,
    ...templateVars,
  })
  return [outputPath, content]
}

/**
 * Reduces boilerplate for creating a yargs handler that writes a component to a
 * location.
 */
// TODO: Make this work for all files, not just components.
export const createYargsForComponentGeneration = ({
  componentName,
  filesFn,
}) => {
  return {
    command: `${componentName} <name>`,
    desc: `Generate a ${componentName} component.`,
    builder: { force: { type: 'boolean', default: false } },
    handler: async ({ force, ...rest }) => {
      const tasks = new Listr(
        [
          {
            title: `Generating ${componentName} files...`,
            task: async () => {
              const f = await filesFn(rest)
              return writeFilesTask(f, { overwriteExisting: force })
            },
          },
        ],
        { collapse: false }
      )

      try {
        await tasks.run()
      } catch (e) {
        // do nothing
      }
    },
  }
}
