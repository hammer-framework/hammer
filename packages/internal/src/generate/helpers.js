import path from 'path'

import { getProject } from '@redwoodjs/structure'

const project = getProject()

export function generateTypeDef(filename, templatePath) {
  const { host } = project
  host.writeFileSync(
    path.join(host.paths.types, filename),
    project.host.readFileSync(templatePath)
  )
}
