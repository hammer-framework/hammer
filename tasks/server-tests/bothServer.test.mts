import { fileURLToPath } from 'node:url'

import { afterAll, afterEach, beforeAll, describe, expect, it } from 'vitest'
import { fs, path, $ } from 'zx'

import { getConfig } from '@redwoodjs/project-config'
import { sleep } from './helpers.mjs'

//////////
// Setup
//////////

$.verbose = !!process.env.VERBOSE

const __dirname = fileURLToPath(new URL('./', import.meta.url))
const FIXTURE_PATH = fileURLToPath(
  new URL('./fixtures/redwood-app', import.meta.url)
)

// @redwoodjs/cli (yarn rw)
const rw = path.resolve(__dirname, '../../packages/cli/dist/index.js')
// @redwoodjs/api-server (yarn rw-server)
const rwServer = path.resolve(
  __dirname,
  '../../packages/api-server/dist/bin.js'
)
// @redwoodjs/web-server (yarn rw-web-server)
const rwWebServer = path.resolve(
  __dirname,
  '../../packages/web-server/dist/bin.js'
)

let original_RWJS_CWD
let projectConfig
beforeAll(() => {
  original_RWJS_CWD = process.env.RWJS_CWD
  process.env.RWJS_CWD = FIXTURE_PATH
  projectConfig = getConfig()
  console.log([
    'These tests use the following command to run the server:',
    `• RWJS_CWD=${process.env.RWJS_CWD} yarn node ${rw} serve`,
    `• RWJS_CWD=${process.env.RWJS_CWD} yarn node ${rwServer}`,
    `• RWJS_CWD=${process.env.RWJS_CWD} yarn node ${rwWebServer}`,
  ].join('\n'))
})
afterAll(() => {
  process.env.RWJS_CWD = original_RWJS_CWD
})

// Clean up the child process after each test
let p
afterEach(async () => {
  if (!p) {
    return
  }
  p.kill()
  // Wait for child process to terminate
  try {
    await p
  } catch {
    // Ignore
  }
})

const TIMEOUT = 1_000 * 2

//////////
// Tests
//////////

describe.each([
  [[rw, 'serve']],
  [rwServer],
])('serve both (%s)', (cmd) => {
  it("has help configured", async () => {
    const { stdout } = await $`yarn node ${cmd} --help`
    expect(stdout).toMatchSnapshot()
  })

  it('errors out on unknown args', async () => {
    try {
      await $`yarn node ${cmd} --foo --bar --baz`
      expect(true).toEqual(false)
    } catch (p) {
      expect(p.exitCode).toEqual(1)
      expect(p.stdout).toEqual('')
      expect(p.stderr).toMatchSnapshot()
    }
  })

  describe('webPort', () => {
    it("`--webPort` changes the web server's port", async () => {
      const webPort = 8920
      p = $`yarn node ${cmd} --webPort ${webPort}`
      await sleep(TIMEOUT)
      await test({ webPort })
    })

    it("`REDWOOD_WEB_PORT` changes the web server's port", async () => {
      process.env.REDWOOD_WEB_PORT = '8921'
      const webPort = +process.env.REDWOOD_WEB_PORT
      p = $`yarn node ${cmd}`
      await sleep(TIMEOUT)
      await test({ webPort })
      delete process.env.REDWOOD_WEB_PORT
    })

    it('`--webPort` takes precedence over `REDWOOD_WEB_PORT`', async () => {
      const webPort = 8922
      process.env.REDWOOD_WEB_PORT = '8923'
      p = $`yarn node ${cmd} --webPort ${webPort}`
      await sleep(TIMEOUT)
      await test({ webPort })
      delete process.env.REDWOOD_WEB_PORT
    })

    it('`[web].port` changes the port', async () => {
      p = $`yarn node ${cmd}`
      await sleep(TIMEOUT)
      await test({ port: projectConfig.web.port })
    })
  })

  describe('webHost', () => {
    it("`--webHost` changes the web server's host", async () => {
      const webHost = '127.0.0.1'
      p = $`yarn node ${cmd} --webHost ${webHost}`
      await sleep(TIMEOUT)
      await test({ webHost })
    })

    it("`REDWOOD_WEB_HOST` changes the web server's host", async () => {
      process.env.REDWOOD_WEB_HOST = '::1'
      const webHost = process.env.REDWOOD_WEB_HOST
      p = $`yarn node ${cmd}`
      await sleep(TIMEOUT)
      await test({ webHost })
      delete process.env.REDWOOD_WEB_HOST
    })

    it('`--webHost` takes precedence over `REDWOOD_WEB_HOST`', async () => {
      const webHost = '::'
      process.env.REDWOOD_WEB_HOST = '0.0.0.0'
      p = $`yarn node ${cmd} --webHost ${webHost}`
      await sleep(TIMEOUT)
      await test({ webHost })
      delete process.env.REDWOOD_WEB_HOST
    })

    it.todo('`[web].host` changes the host')

    it("defaults to '::' if `NODE_ENV` isn't production", async () => {
      p = $`yarn node ${cmd}`
      await sleep(TIMEOUT)
      await test()
    })

    it.todo("defaults to '0.0.0.0' if `NODE_ENV` is production")
  })
})

async function test(options = {}) {
  options.webHost ??= '::'
  if (options.webHost.includes(':')) {
    options.webHost = `[${options.webHost}]`
  }
  options.webPort ??= projectConfig.web.port

  const webRes = await fetch(`http://${options.webHost}:${options.webPort}/about`)
  const webBody = await webRes.text()

  expect(webRes.status).toEqual(200)
  expect(webBody).toEqual(
    fs.readFileSync(
      path.join(__dirname, './fixtures/redwood-app/web/dist/about.html'),
      'utf-8'
    )
  )

  options.apiHost ??= '::'
  if (options.apiHost.includes(':')) {
    options.apiHost = `[${options.apiHost}]`
  }
  options.apiPort ??= projectConfig.api.port

  const apiRes = await fetch(`http://${options.apiHost}:${options.apiPort}/hello`)
  const apiBody = await apiRes.json()

  expect(apiRes.status).toEqual(200)
  expect(apiBody).toEqual({ data: 'hello function' })
}
