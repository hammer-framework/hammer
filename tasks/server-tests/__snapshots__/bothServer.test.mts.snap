// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`serve both (/Users/dom/projects/redwood/redwood/packages/api-server/dist/bin.js) > errors out on unknown args 1`] = `
"rw-server

Start a server for serving the api and web sides

Commands:
  rw-server      Start a server for serving the api and web sides      [default]
  rw-server api  Start a server for serving the api side
  rw-server web  Start a server for serving the web side

Options:
      --webPort, --web-port                 The port for the web server to
                                            listen on                   [number]
      --webHost, --web-host                 The host for the web server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiPort, --api-port                 The port for the api server to
                                            listen on                   [number]
      --apiHost, --api-host                 The host for the api server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiRootPath, --api-root-path,       Root path where your api functions
      --rootPath, --root-path               are served   [string] [default: "/"]
  -h, --help                                Show help                  [boolean]
  -v, --version                             Show version number        [boolean]

Unknown arguments: foo, bar, baz
"
`;

exports[`serve both (/Users/dom/projects/redwood/redwood/packages/api-server/dist/bin.js) > has help configured 1`] = `
"rw-server

Start a server for serving the api and web sides

Commands:
  rw-server      Start a server for serving the api and web sides      [default]
  rw-server api  Start a server for serving the api side
  rw-server web  Start a server for serving the web side

Options:
      --webPort, --web-port                 The port for the web server to
                                            listen on                   [number]
      --webHost, --web-host                 The host for the web server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiPort, --api-port                 The port for the api server to
                                            listen on                   [number]
      --apiHost, --api-host                 The host for the api server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiRootPath, --api-root-path,       Root path where your api functions
      --rootPath, --root-path               are served   [string] [default: "/"]
  -h, --help                                Show help                  [boolean]
  -v, --version                             Show version number        [boolean]
"
`;

exports[`serve both ([ '/Users/dom/projects/redwood/redwood/packages/cli/dist/index.js', 'serve' ]) > errors out on unknown args 1`] = `
"rw serve [side]

Start a server for serving both the api and web sides

Commands:
  rw serve      Start a server for serving the api and web sides       [default]
  rw serve api  Start a server for serving the api side
  rw serve web  Start a server for serving the web side

Options:
      --version                             Show version number        [boolean]
      --cwd                                 Working directory to use (where
                                            \`redwood.toml\` is located)
      --telemetry                           Whether to send anonymous usage
                                            telemetry to RedwoodJS     [boolean]
      --webPort, --web-port                 The port for the web server to
                                            listen on                   [number]
      --webHost, --web-host                 The host for the web server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiPort, --api-port                 The port for the api server to
                                            listen on                   [number]
      --apiHost, --api-host                 The host for the api server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiRootPath, --api-root-path,       Root path where your api functions
      --rootPath, --root-path               are served   [string] [default: "/"]
  -h, --help                                Show help                  [boolean]

Also see the Redwood CLI Reference
(​https://redwoodjs.com/docs/cli-commands#serve​)

Unknown arguments: foo, bar, baz
"
`;

exports[`serve both ([ '/Users/dom/projects/redwood/redwood/packages/cli/dist/index.js', 'serve' ]) > has help configured 1`] = `
"rw serve [side]

Start a server for serving both the api and web sides

Commands:
  rw serve      Start a server for serving the api and web sides       [default]
  rw serve api  Start a server for serving the api side
  rw serve web  Start a server for serving the web side

Options:
      --version                             Show version number        [boolean]
      --cwd                                 Working directory to use (where
                                            \`redwood.toml\` is located)
      --telemetry                           Whether to send anonymous usage
                                            telemetry to RedwoodJS     [boolean]
      --webPort, --web-port                 The port for the web server to
                                            listen on                   [number]
      --webHost, --web-host                 The host for the web server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiPort, --api-port                 The port for the api server to
                                            listen on                   [number]
      --apiHost, --api-host                 The host for the api server to
                                            listen on. Note that you most likely
                                            want this to be '0.0.0.0' in
                                            production                  [string]
      --apiRootPath, --api-root-path,       Root path where your api functions
      --rootPath, --root-path               are served   [string] [default: "/"]
  -h, --help                                Show help                  [boolean]

Also see the Redwood CLI Reference
(​https://redwoodjs.com/docs/cli-commands#serve​)
"
`;
